src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -747,50 +758,61 @@ private Task WriteSuffix()
             bool appCompleted)
         {
             var responseHeaders = FrameResponseHeaders;
-            responseHeaders.SetReadOnly();
-
             var hasConnection = responseHeaders.HasConnection;
 
             var end = SocketOutput.ProducingStart();
+
             if (_keepAlive && hasConnection)
             {
                 var connectionValue = responseHeaders.HeaderConnection.ToString();
                 _keepAlive = connectionValue.Equals("keep-alive", StringComparison.OrdinalIgnoreCase);
             }
 
-            if (!responseHeaders.HasTransferEncoding && !responseHeaders.HasContentLength)
+            // Set whether response can have body
+            _canHaveBody = StatusCanHaveBody(StatusCode) && Method != "HEAD";
```

Comment:
Since 101 status can have a body of sorts, I would prefer `_canHaveBody` is true in that case. When deciding whether or not to set `Content-Length: 0` we can check for 101 specifically.